Avoid rollback after a commit failure in `TransactionalOperator`

EnricSala opened the pull request:
A failure to commit a reactive transaction will complete the transaction and clean up resources. Executing a rollback at that point is invalid, which causes an `IllegalTransactionStateException` that masks the cause of the commit failure.

This change restructures `TransactionalOperatorImpl` and `ReactiveTransactionSupport` to avoid executing a rollback after a failed commit. While there, the `Mono` transaction handling in `TransactionalOperator` is simplified by moving it to a default method on the interface.

See gh-27523

EnricSala committed:
Option 1: unwrap the `RuntimeException` from `usingWhen`

EnricSala reviewed (commented):
Left a couple of comments to explain some changes.

EnricSala committed:
Option 2: drop the unwrapping, propagate `RuntimeException`

EnricSala committed:
Revert "Option 2: drop the unwrapping, propagate `RuntimeException`"

This reverts commit aaad05147b3c6fb820ab6a095454aee1f9c70f12.

EnricSala commented on spring-tx/src/main/java/org/springframework/transaction/reactive/TransactionalOperatorImpl.java:
```diff
@@ -70,40 +71,19 @@ public ReactiveTransactionManager getTransactionManager() {
 	public <T> Flux<T> execute(TransactionCallback<T> action) throws TransactionException {
-		return TransactionContextManager.currentContext().flatMapMany(context -> {
-			Mono<ReactiveTransaction> status = this.transactionManager.getReactiveTransaction(this.transactionDefinition);
-			// This is an around advice: Invoke the next interceptor in the chain.
-			// This will normally result in a target object being invoked.
-			// Need re-wrapping of ReactiveTransaction until we get hold of the exception
-			// through usingWhen.
-			return status.flatMapMany(it -> Flux
-					.usingWhen(
-							Mono.just(it),
```
I would also have a preference for this option because the `RuntimeException` feels a bit synthetic, it's only there due to the `usingWhen` implementation, and it doesn't align with the behavior of non-reactive transactions.

I have applied the change, it should be ready for review :)

sdeleuze commented:
@mp911de Could you please review the latest version of this PR and confirm (or not) you are ok for merging it (I will also have a deeper look after your confirmation)?

sdeleuze commented:
@EnricSala I have rebased the branch on top of `main` in https://github.com/sdeleuze/spring-framework/tree/gh-27523, but I see an error in `CoroutinesTransactionInterceptorTests`, could you please have a look and maybe rebased this PR on top of `main` potentially using my branch as a basis?

EnricSala committed:
Fix test `CoroutinesTransactionInterceptorTests`

EnricSala commented:
@sdeleuze I have rebased this PR on top of `main` and added a commit to fix `CoroutinesTransactionInterceptorTests`. Please check if this small fix makes sense :)

sdeleuze commented:
Merged via edf0ae77e502dd0e5b85a4fa8fc08d3e516d990a after @simonbasle and @mp911de green light. Thanks for contributing this and for your patience @EnricSala. Please test snapshots to check everything looks fine for your use cases.
